### ISSUE TITLE
CHUNK 6 OF 6 — DEPLOYMENT, CI/CD & FINAL INTEGRATION

### ISSUE THREAD

**Issue by SunbalAzizLCWU** [SunbalAzizLCWU/medwiz#13]
You are completing the AuraNode build. This final chunk wires 
everything together, adds deployment configs, and ensures the 
entire system is production-ready end to end.

================================================================
PART A — FRONTEND CI/CD
================================================================

----------------------------------------------------------------
FILE: auranode-web/.github/workflows/ci.yml
----------------------------------------------------------------
Complete GitHub Actions workflow:
name: Frontend CI
Triggers: push and pull_request on main and dev

Jobs:
  build-and-lint:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v4
      - uses: actions/setup-node@v4
        with: { node-version: '20', cache: 'npm' }
      - run: npm ci
      - run: npm run lint
      - run: npm run build
    env:
      NEXT_PUBLIC_SUPABASE_URL: https://mock.supabase.co
      NEXT_PUBLIC_SUPABASE_ANON_KEY: mock_anon_key
      NEXT_PUBLIC_API_URL: http://localhost:8000
      NEXT_PUBLIC_APP_URL: http://localhost:3000

----------------------------------------------------------------
FILE: auranode-web/vercel.json
----------------------------------------------------------------
{
  "framework": "nextjs",
  "buildCommand": "npm run build",
  "outputDirectory": ".next",
  "rewrites": [
    {
      "source": "/api/backend/:path*",
      "destination": "YOUR_RENDER_URL/api/:path*"
    }
  ]
}
Note: user will replace YOUR_RENDER_URL with actual Render URL.

================================================================
PART B — INTEGRATION: END-TO-END API CLIENT
================================================================

----------------------------------------------------------------
FILE: auranode-web/lib/api-client.ts
----------------------------------------------------------------
Complete typed API client for all backend calls:

const API_URL = process.env.NEXT_PUBLIC_API_URL;

async function getAuthHeaders(): Promise<Record<string, string>> {
  const supabase = createClient();
  const { data: { session } } = await supabase.auth.getSession();
  if (!session) throw new Error("Not authenticated");
  return {
    "Content-Type": "application/json",
    "Authorization": `Bearer ${session.access_token}`,
  };
}

async function handleResponse<T>(res: Response): Promise<T> {
  if (!res.ok) {
    const error = await res.json().catch(() => ({}));
    throw new Error(error.detail || `HTTP ${res.status}`);
  }
  return res.json();
}

Export these typed functions:

export const apiClient = {
  // Reports
  createReport: async (data: CreateReportPayload): 
    Promise<ReportResponse> => { ... },
  
  getReports: async (status?: ReportStatus): 
    Promise<ReportResponse[]> => { ... },
  
  getReport: async (id: string): 
    Promise<ReportResponse> => { ... },
  
  approveReport: async (id: string, doctor_notes: string): 
    Promise<{ success: boolean; signed_url_token: string }> => 
    { ... },
  
  getPublicReport: async (token: string): 
    Promise<Partial<ReportResponse>> => {
      // No auth header for this one
      const res = await fetch(`${API_URL}/api/reports/public/${token}`);
      return handleResponse(res);
    },

  retryReport: async (id: string): 
    Promise<{ message: string }> => { ... },

  // Org
  getDashboardStats: async (orgId: string): 
    Promise<DashboardStats> => { ... },

  inviteDoctor: async (orgId: string, doctorUserId: string): 
    Promise<{ success: boolean }> => { ... },

  // Auth
  verifyPmdc: async (pmdc_number: string): 
    Promise<{ success: boolean }> => { ... },
  
  getMe: async (): 
    Promise<UserResponse> => { ... },
};

Include TypeScript interfaces:
interface CreateReportPayload {
  patient_name: string;
  patient_phone: string;
  upload_type: UploadType;
  file_url: string;
}

================================================================
PART C — BACKEND FINAL WIRING
================================================================

----------------------------------------------------------------
FILE: auranode-api/app/main.py — FINAL VERSION
----------------------------------------------------------------
Complete main.py with everything wired correctly:

from contextlib import asynccontextmanager
from fastapi import FastAPI
from fastapi.middleware.cors import CORSMiddleware
from slowapi import Limiter, _rate_limit_exceeded_handler
from slowapi.util import get_remote_address
from slowapi.errors import RateLimitExceeded

from app.core.config import settings
from app.core.logging_config import setup_logging
from app.workers.job_queue import scheduler
from app.api import health, auth, reports, analyze, org

limiter = Limiter(key_func=get_remote_address)

@asynccontextmanager
async def lifespan(app: FastAPI):
  setup_logging()
  scheduler.start()
  logger.info("AuraNode API started. Scheduler running.")
  yield
  scheduler.shutdown()
  logger.info("AuraNode API shutting down.")

app = FastAPI(
  title="AuraNode API",
  version="1.0.0",
  description="B2B Tele-Diagnostic SaaS Platform",
  lifespan=lifespan
)

app.state.limiter = limiter
app.add_exception_handler(RateLimitExceeded, 
  _rate_limit_exceeded_handler)

app.add_middleware(
  CORSMiddleware,
  allow_origins=settings.ALLOWED_ORIGINS,
  allow_credentials=True,
  allow_methods=["*"],
  allow_headers=["*"],
)

app.include_router(health.router, prefix="/api/health", 
  tags=["health"])
app.include_router(auth.router, prefix="/api/auth", 
  tags=["auth"])
app.include_router(reports.router, prefix="/api/reports", 
  tags=["reports"])
app.include_router(analyze.router, prefix="/api/analyze", 
  tags=["analyze"])
app.include_router(org.router, prefix="/api/org", 
  tags=["org"])

@app.get("/")
async def root():
  return {"message": "AuraNode API", "status": "running", 
          "version": "1.0.0"}

================================================================
PART D — GOOGLE COLAB TRAINING NOTEBOOK
================================================================

----------------------------------------------------------------
FILE: auranode-api/notebooks/train_efficientnet.ipynb
----------------------------------------------------------------
Generate a complete, runnable Jupyter notebook with these cells:

CELL 1 — Install dependencies:
!pip install kaggle torch torchvision timm onnx onnxruntime 
!pip install scikit-learn matplotlib seaborn

CELL 2 — Kaggle Auth & Dataset Download:
# Upload kaggle.json first using: 
# from google.colab import files; files.upload()
import os
os.makedirs("/root/.kaggle", exist_ok=True)
# Copy uploaded kaggle.json to ~/.kaggle/kaggle.json
!cp kaggle.json /root/.kaggle/kaggle.json
!chmod 600 /root/.kaggle/kaggle.json
!kaggle datasets download -d nih-chest-xrays/data 
  --path /content/nih_data -p
# Note: MURA can be added similarly

CELL 3 — Dataset Preparation:
Complete code to:
- Load NIH chest x-ray dataset
- Binary classification: Normal (class 0) vs 
  Abnormal (any pathology = class 1)
- Split: 80% train, 10% val, 10% test
- PyTorch Dataset class with transforms:
  Train: RandomHorizontalFlip, RandomRotation(10), 
         Resize(224,224), ToTensor, Normalize(imagenet)
  Val/Test: Resize(224,224), ToTensor, Normalize(imagenet)
- DataLoaders: batch_size=32, num_workers=2

CELL 4 — Model Definition:
import timm
model = timm.create_model(
  "efficientnet_b0", 
  pretrained=True, 
  num_classes=2
)
# Freeze base layers first
for param in model.parameters():
  param.requires_grad = False
# Unfreeze classifier head
for param in model.classifier.parameters():
  param.requires_grad = True
model = model.cuda()

CELL 5 — Training Phase 1 (head only, 5 epochs):
optimizer = torch.optim.Adam(
  filter(lambda p: p.requires_grad, model.parameters()), 
  lr=1e-3
)
criterion = torch.nn.CrossEntropyLoss()
scheduler = torch.optim.lr_scheduler.CosineAnnealingLR(
  optimizer, T_max=5
)
Complete training loop:
  for epoch in range(5):
    model.train()
    for images, labels in train_loader:
      images, labels = images.cuda(), labels.cuda()
      optimizer.zero_grad()
      outputs = model(images)
      loss = criterion(outputs, labels)
      loss.backward()
      optimizer.step()
    scheduler.step()
    # Validation loop + print val accuracy

CELL 6 — Training Phase 2 (fine-tune all layers, 10 epochs):
# Unfreeze all layers
for param in model.parameters():
  param.requires_grad = True
optimizer = torch.optim.Adam(model.parameters(), lr=1e-4)
scheduler = torch.optim.lr_scheduler.CosineAnnealingLR(
  optimizer, T_max=10
)
(Same training loop, 10 epochs)

CELL 7 — Evaluation:
from sklearn.metrics import roc_auc_score, classification_report
Evaluate on test set:
- Print classification_report
- Calculate and print AUC-ROC score
- Plot confusion matrix using seaborn
- Plot ROC curve using matplotlib
Target: AUC-ROC >= 0.75 for export

CELL 8 — ONNX Export:
model.eval()
model.cpu()
dummy_input = torch.randn(1, 3, 224, 224)
torch.onnx.export(
  model,
  dummy_input,
  "efficientnet_b0_v1.onnx",
  export_params=True,
  opset_version=11,
  input_names=["input"],
  output_names=["output"],
  dynamic_axes={
    "input": {0: "batch_size"},
    "output": {0: "batch_size"}
  }
)
print("Model exported to efficientnet_b0_v1.onnx")

CELL 9 — Verify ONNX Export:
import onnxruntime as ort
import numpy as np
session = ort.InferenceSession("efficientnet_b0_v1.onnx",
  providers=["CPUExecutionProvider"])
dummy = np.random.randn(1, 3, 224, 224).astype(np.float32)
outputs = session.run(None, {"input": dummy})
print("ONNX model output shape:", outputs[0].shape)
print("ONNX verification PASSED ✓")

CELL 10 — Download the model:
from google.colab import files
files.download("efficientnet_b0_v1.onnx")
print("Download your efficientnet_b0_v1.onnx file now.")
print("Then upload it to Supabase Storage bucket: ml-models")

================================================================
PART E — README FILES
================================================================

----------------------------------------------------------------
FILE: auranode-api/README.md
----------------------------------------------------------------
Complete README with:
# AuraNode API
## Architecture Notes (ONNX, Groq, APScheduler)

----------------------------------------------------------------
FILE: auranode-web/README.md
----------------------------------------------------------------
Complete README with:
# AuraNode Web
## Key Components Guide

================================================================
PART F — FINAL INTEGRATION CHECKLIST
================================================================

Generate a file: LAUNCH_CHECKLIST.md (in both repos root)

Content: Complete pre-launch checklist:

## Supabase Setup
- [ ] Project created at supabase.com
- [ ] SQL schema executed (supabase/schema.sql)
- [ ] RLS enabled on all 7 tables
- [ ] Storage bucket "medical_scans" created (private)
- [ ] Storage bucket "ml-models" created (public)
- [ ] ONNX model uploaded to ml-models bucket
- [ ] Supabase Realtime enabled for reports and notifications tables
- [ ] JWT secret copied from Supabase Settings → API

## Backend (Render)
- [ ] GitHub repo connected to Render
- [ ] All environment variables set in Render dashboard
- [ ] Health check path set to /api/health
- [ ] Deployment successful (check logs)
- [ ] Test: curl https://your-api.onrender.com/api/health

## Frontend (Vercel)
- [ ] GitHub repo connected to Vercel
- [ ] All NEXT_PUBLIC_ environment variables set
- [ ] NEXT_PUBLIC_API_URL set to Render URL
- [ ] Deployment successful
- [ ] Test: visit https://your-app.vercel.app

## UptimeRobot
- [ ] Free account created at uptimerobot.com
- [ ] HTTP monitor added: https://your-api.onrender.com/api/health
- [ ] Check interval: 14 minutes
- [ ] Alert email configured

## Sentry (Error Monitoring)
- [ ] Free project at sentry.io
- [ ] Backend: pip install sentry-sdk, add DSN to .env
  Add to main.py: 
  import sentry_sdk
  sentry_sdk.init(dsn=os.getenv("SENTRY_DSN"), 
                   traces_sample_rate=0.1)
- [ ] Frontend: npm install @sentry/nextjs, run npx @sentry/wizard

## ML Model
- [ ] Colab notebook run successfully
- [ ] AUC-ROC >= 0.75 achieved
- [ ] ONNX file downloaded from Colab
- [ ] ONNX file uploaded to Supabase ml-models bucket
- [ ] ONNX_MODEL_URL env var set in Render

## End-to-End Test
- [ ] Create clinic_admin account via /signup
- [ ] Create organization (via superadmin or directly in Supabase)
- [ ] Create doctor account, verify PMDC
- [ ] Upload a test X-ray image
- [ ] Verify report appears in dashboard
- [ ] Verify AI processing completes
- [ ] Log in as doctor, approve report
- [ ] Verify WhatsApp button generates correct link
- [ ] Verify PDF downloads correctly
- [ ] Verify public report link works

After generating all files, output:
"CHUNK 6 COMPLETE — AuraNode is fully built. 
Entire system from 0 to 100% complete.
Follow the LAUNCH_CHECKLIST.md to go live."
